### PR TITLE
Site migration: vertical centre progress view on larger viewports

### PR DESCRIPTION
**Pull request by andfinally**
Note: This change is behind a feature flag and is part of an in-progress larger project. For context, see pbkcP4-8-p2.

#### Changes proposed in this Pull Request

* Vertically centres the migration progress view on viewports 660px and up.

image

* I've only done the vertical centring on larger viewports, as on smaller ones there's more of a danger the content will bump into the Masterbar if it ever grows:

image

#### Testing instructions

* Check out this pull request and run Calypso locally, or view it on calypso.live with the flag tools/migrate.
* Viewing calypso.localhost:3000/migrate/, you should see a SiteSelector component allowing you to choose a site to migrate a Jetpack site into.
* When you select a site, you should see another SiteSelector component, listing all the Jetpack sites your account has access to.
* Selecting one of those sites should take you to a confirmation screen.
* Accept the confirmation. You should see a progress screen detailing the steps of the migration. The content should be vertically centred on viewports larger than 660px.
* After the migration ends, you should be able to scroll down Calypso pages as normal. (It helps if you switch to a mobile view in your browser's dev tools to check this.)

### CODE REVIEW

**Comment by andfinally** [comment]
Hi @danhauk - the only way I could find to make the parent containers stretch to full height was to give them `height: 100%`. Unfortunately this meant content wasn't scrolling on mobile, so I'm only doing this on larger viewports while the progress view is displayed. I don't know if you can think of a better way to do this!

**Review by mattsherman — CHANGES_REQUESTED** [review]
I did not test this.

Looking at the code, this feels like too complex of a solution for what we are aiming to do. We should be able to vertically center content via CSS alone. 

@danhauk Can you please take a look at this and provide us with some suggestions on how to vertically center this without Redux state actions and such? CSS only.

**Comment by andfinally** [comment]
As far as I can see it's not possible to centre the progress view vertically without touching the state, because the component's parent containers are no taller than the component itself. We need to tell them to stretch out to full height, which involves a state change to add a class to the `layout` div.

File: client/state/action-types.js [hunk 1]
```diff
@@ -284,6 +284,8 @@ export const FOLLOWERS_REQUEST_ERROR = 'FOLLOWERS_REQUEST_ERROR';
 export const FOLLOWER_REMOVE_ERROR = 'FOLLOWER_REMOVE_ERROR';
 export const FOLLOWER_REMOVE_REQUEST = 'FOLLOWER_REMOVE_REQUEST';
 export const FOLLOWER_REMOVE_SUCCESS = 'FOLLOWER_REMOVE_SUCCESS';
+export const FULL_HEIGHT_LAYOUT_ON = 'FULL_HEIGHT_LAYOUT_ON';
+export const FULL_HEIGHT_LAYOUT_OFF = 'FULL_HEIGHT_LAYOUT_OFF';
```
Comment (jsnajdr) [inline client/state/action-types.js]: There can be just one action, `FULL_HEIGHT_LAYOUT_TOGGLE`. No need to spam the already long list of Redux actions 🙂

File: client/state/ui/full-height-layout/actions.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+ * Internal dependencies
+ */
+import { FULL_HEIGHT_LAYOUT_ON, FULL_HEIGHT_LAYOUT_OFF } from 'state/action-types';
+
+export function setFullHeightLayout() {
+	return dispatch =>
+		dispatch( {
+			type: FULL_HEIGHT_LAYOUT_ON,
+		} );
+}
```
Comment (jsnajdr) [inline client/state/ui/full-height-layout/actions.js]: Why is this action a thunk?

Both actions can be merged into one with a boolean parameter.

**Review by jsnajdr — COMMENTED** [review]
Vertically centering a section content in Calypso indeed seems to be harder than expected 🙂 

I think that the root cause is that the Calypso layout is designed for the content to be document-like, i.e., its height is determined by the content. And the content height is not related to the browser viewport size -- it scrolls when it overflows and doesn't fill the whole content are when it's smaller.

Vertical centering is not a well-defined operation in such layout. Personally I think it's not a good idea in this case (migration progress screen), but it's not up to me to make the decision here -- instead, I'll help you implement _your_ decision 🙂 

If we want to center the migration progress UI, we indeed need to hack the parent layout styles a bit. And we currently customize the `<Layout>` component through Redux, although I'm also not very happy about it.

There's some prior art for customizing `<Layout>` with Redux and the section configuration, namely the "has masterbar" and "has sidebar" flags. Sections and screens (e.g., Jetpack Connect) customize that by either having a `secondary: false` flag in its section definition, or by dispatching Redux actions like `hideMasterbar` in their page.js handlers.

I'm proposing some reduction of Redux verbosity in my review comments.

Also, dispatching Redux actions in `render` methods is a big ⛔️. Instead, a very elegant solution is to create a helper component that doesn't render anything, but sets and unsets the `full-height-layout` flag on mount and unmount:
```jsx
function FullHeightLayout() {
  const dispatch = useDispatch();
  useEffect( () => {
    dispatch( setFullHeightLayout( true ) );
    return () => {
      dispatch( setFullHeightLayout( false ) );
    }
  }, [ dispatch ] );

  return null;
}
```

And then you can use it to render the migration progress:
```jsx
renderMigrationProgress() {
  return (
    <Card className="migrate__pane">
      <FullHeightLayout />
      { ... }
    </Card>
  );
}
```
The full-height-layout flag will then be automatically set and reset as the migration progress UI is mounted and unmounted.

File: client/my-sites/migrate/section-migrate.scss [hunk 3]
```diff
@@ -1,3 +1,17 @@
+$masterbar-height: 46px;
+
+@include breakpoint( '>660px' ) {
```
Comment (jsnajdr) [inline client/my-sites/migrate/section-migrate.scss]: I think what you really want to do here is to do vertical centering only if the layout _height_ is large enough. But this condition applies to the layout _width_. The width and height are not related to each other -- only loosely correlated.

Using `@media ( max-height )` or `@media ( min-height )` seems like a better option.

File: client/my-sites/migrate/section-migrate.jsx [hunk 4]
```diff
@@ -206,6 +212,8 @@ class SectionMigrate extends Component {
 	}
 
 	renderMigrationError() {
+		this.props.unsetFullHeightLayout();
```
[the remark this answers is not in the document]
Comment (andfinally) [reply client/my-sites/migrate/section-migrate.jsx]: 😬 Ow! Thanks.

File: client/my-sites/migrate/section-migrate.scss [hunk 5]
```diff
@@ -1,3 +1,17 @@
...
+	.layout.is-section-migrate.is-full-height-layout {
```
[the remark this answers is not in the document]
Comment (andfinally) [reply client/my-sites/migrate/section-migrate.scss]: My motivation here was concern not to break scrolling in any other section.

[the remark this answers is not in the document]
Comment (jsnajdr) [reply client/my-sites/migrate/section-migrate.scss]: Just the `.layout.is-full-height-layout` selector is already specific enough. And is guaranteed to apply only if you enable the full height layout, i.e., in only when the "migration progress" UI is shown.

File: client/my-sites/migrate/section-migrate.jsx [hunk 6]
```diff
@@ -229,8 +237,10 @@ class SectionMigrate extends Component {
 		const sourceSiteDomain = get( sourceSite, 'domain' );
 		const targetSiteDomain = get( targetSite, 'domain' );
 
+		this.props.setFullHeightLayout();
+
 		return (
-			<>
+			<div className="migrate__wrapper">
```
[the remark this answers is not in the document]
Comment (andfinally) [reply client/my-sites/migrate/section-migrate.jsx]: Thanks, I'll try it without the extra div. You're right, the original fragment was unnecessary too.

File: client/my-sites/migrate/section-migrate.scss (hunk 3 above)
Comment (andfinally) [reply client/my-sites/migrate/section-migrate.scss]: I was relying on the loose correlation, but that would probably be better.

File: client/my-sites/migrate/section-migrate.scss [hunk 7]
```diff
@@ -72,12 +87,22 @@
...
+
 .migrate__pane {
 	display: flex;
 	flex-direction: column;
 	align-items: center;
 	background: inherit;
 	box-shadow: none;
+	@include breakpoint( '>660px' ) {
+		margin-top: -$masterbar-height;
+	}
```
[the remark this answers is not in the document]
Comment (andfinally) [reply client/my-sites/migrate/section-migrate.scss]: That's right – the full window centring looked better to me, but maybe you're right.

File: client/state/ui/full-height-layout/actions.js (hunk 2 above)
Comment (andfinally) [reply client/state/ui/full-height-layout/actions.js]: This is another case where my React understanding wasn't up to scratch.

**Comment by andfinally** [comment]
Thanks very much @jsnajdr! 🙌 After some discussion it seems like it this layout not be worth the extra Redux complication it adds – I'm hoping we can go ahead without it. But I really appreciate your thoughtful and informative review. It's helped expose some gaps in my knowledge, which I'll try to fill.

**Comment by jsnajdr** [comment]
Thanks @andfinally for uncovering some design deficiencies and inflexibilities in the Calypso layout design! 🤝

**Comment by andfinally** [comment]
Closing this one till we can figure out if we do want to vertically centre the progress content.

**Comment by danhauk** [comment]
@andfinally I took a look at this and tried a few things with just CSS. But I think you're right that it's dependent on state to get the 100% height. I'm ok moving forward without vertically centering the content.

**Comment by mattsherman** [comment]
Thanks @jsnajdr for the detailed review on this a few days ago. I really appreciate it.

**Pull request closed without merging**